Implement browser history search feature

hunarjain09 opened the pull request:
Summary

  This PR significantly enhances the search-all-tabs extension by adding browser history search capabilities and a filter UI to
  distinguish between active tabs and history items. The extension now searches through both active tabs and browser history, providing
  users with comprehensive search functionality across their browsing experience.

  Key Features Added

  🕒 Browser History Search

  - Full-text search through browser history using the same Xapian search engine
  - Content fetching via HTTP requests to parse HTML content from history URLs
  - Configurable time range (default: 7 days) and result limits (default: 1000 items)
  - Visual distinction between history items and active tabs in search results

  🔗 Filter UI Enhancement

  - Smart filter button in popup header with dropdown menu
  - Three filter options: All Results, Active Tabs, Browser History
  - Real-time result counts for each category
  - Visual indicators: 🔗 All, 📑 Tabs, 🕒 History

  📊 Enhanced Debug Logging

  - Stack-based logging with 2KB memory limit
  - Persistent file logging with downloadable debug files
  - Compact log format optimized for storage efficiency
  - Context-aware logging (POPUP, OPTIONS, BACKGROUND, etc.)

  Technical Implementation

  Core Changes

  1. Manifest Updates (manifest.json)

  "permissions": [
    "history",      // NEW: Access browser history
    "downloads"     // NEW: Download debug logs
  ],
  "version": "0.6.4"

  2. History Search Engine (indexer.js)

  - queryHistory(): Fetches browser history items with time-based filtering
  - inspectHistory(): HTTP content fetching and HTML parsing for history items
  - Parallel processing: Active tabs and history items processed simultaneously
  - Content extraction: Title, description, keywords, body text from HTML

  3. Enhanced Result Handling (index.js)

  - TabId management: History items use tabId = -1 for proper identification
  - Click handling: History items open in new tabs, active tabs get focused
  - Visual indicators: Blue border and clock icon for history items
  - Error handling: Comprehensive logging for debugging tab/history issues

  4. Background Worker Validation (worker.js)

  // Prevent script injection on history items
  if (request.tabId <= 0) {
    console.debug('⚠️ Ignoring find request for invalid tabId:', request.tabId);
    return true;
  }

  5. Stack-Based Logger (logger.js)

  // LIFO stack with automatic trimming
  while (this.currentLogSize > this.maxLogSize || this.logs.length > this.maxEntries) {
    const removedEntry = this.logs.shift(); // Remove oldest
    if (removedEntry) this.currentLogSize -= removedEntry.length;
  }

  Filter UI Implementation

  HTML Structure (index.html)

  <div id="filter-controls">
    <button id="filter-toggle">
      <span id="filter-icon">🔗</span>
      <span id="filter-text">All</span>
      <span id="filter-arrow">▼</span>
    </button>
    <div id="filter-menu" class="hidden">
      <div class="filter-option" data-filter="all">
        <span class="filter-icon">🔗</span>
        <span>All Results</span>
        <span class="filter-count" id="count-all"></span>
      </div>
      <!-- More options... -->
    </div>
  </div>

  CSS Styling (index.css)

  - Responsive dropdown with proper z-index (9999) and positioning
  - Hover states and active state indicators
  - Visual distinction for history items (blue border, clock icon)
  - Compact design that fits existing UI seamlessly

  JavaScript Logic (index.js)

  - Event handlers for filter toggle and option selection
  - Result filtering using .filter-hidden CSS class
  - Count updates in real-time as search results change
  - State persistence maintains selected filter across searches

  Problem Resolution

  Critical Fixes Applied

  1. Manifest V3 Compatibility

  - Issue: 'background.scripts' requires manifest version of 2 or lower
  - Fix: Removed deprecated scripts array, using only service_worker

  2. Script Injection Errors

  - Issue: Background worker attempting to inject scripts into history items
  - Root Cause: History items getting string IDs instead of numeric tabId = -1
  - Fix: Explicit item.id = -1 assignment + validation in background worker

  3. History Items Not Opening

  - Issue: History items with null/undefined tabId not triggering new tab creation
  - Fix: Enhanced condition checking: tabIdNumber <= 0 || isNaN(tabIdNumber) || tabId === null

  4. UI Cropping Issues

  - Issue: Filter dropdown menu being cut off
  - Fix: Increased z-index to 9999, added overflow: visible, proper positioning

  Testing Strategy

  Manual Testing Scenarios

  1. Basic functionality: Search active tabs vs history items
  2. Filter operations: Toggle between All/Tabs/History filters
  3. Click behavior: History opens new tab, active tabs get focused
  4. Edge cases: Empty results, network errors, CSP restrictions
  5. Performance: Large history datasets, concurrent searches

  Debug Logging Verification

  - Console logs show clear distinction between tab/history processing
  - File logs downloadable with context-aware information
  - Stack-based logging maintains recent entries within 2KB limit

  Configuration Options

  New Options Added (options/index.html)

  <input type="checkbox" id="history-enabled"> Enable browser history search
  <input type="number" id="history-days" value="7"> History search days
  <input type="number" id="history-max-results" value="1000"> Max history results

  Debug Controls

  - Download logs: Get debug file for troubleshooting
  - Clear logs: Reset debug log stack
  - Log statistics: View utilization and entry counts

  Performance Considerations

  Optimizations Implemented

  - Batch processing: 5-item batches for indexing performance
  - Content length limits: Configurable max content size per item
  - Duplicate filtering: URL-based deduplication across tabs and history
  - Timeout handling: 10-second fetch timeout for history content
  - Memory management: Stack-based logging with automatic cleanup

  Resource Usage

  - Storage: ~2KB for debug logs + indexed content in Xapian database
  - Network: HTTP requests only for history items (user-configurable)
  - CPU: Parallel processing of active tabs and history items

  Backward Compatibility

  Preserved Features

  - ✅ All existing search functionality works unchanged
  - ✅ Existing keyboard shortcuts and UI interactions
  - ✅ Performance characteristics maintained for active tab search
  - ✅ Configuration options remain optional (history disabled by default)

  Migration Path

  - No breaking changes: Extension works immediately after update
  - Progressive enhancement: History search is opt-in via options page
  - Graceful degradation: Works without history permission if denied

  Code Quality

  Architecture Improvements

  - Separation of concerns: History logic isolated in dedicated methods
  - Error handling: Comprehensive try/catch blocks with informative logging
  - Code documentation: Clear variable names and inline comments
  - Consistent patterns: Follows existing codebase conventions

  Security Considerations

  - CORS handling: Proper error handling for cross-origin history requests
  - Permission validation: Checks for history permission before attempting access
  - Content sanitization: HTML parsing through DOMParser with safety checks

  Files Modified

  Core Extension Files

  - manifest.json - Added permissions, version bump
  - data/popup/indexer.js - History search implementation
  - data/popup/index.js - UI enhancements and click handling
  - data/popup/index.html - Filter controls HTML structure
  - data/popup/index.css - Filter UI styling and history indicators
  - worker.js - Background script validation enhancements

  New/Enhanced Files

  - data/logger.js - Stack-based logging system
  - data/options/index.html - History configuration options

  Future Enhancements

  Potential Improvements

  - Smart history filtering: ML-based relevance scoring for history items
  - Bookmark search: Extend to search browser bookmarks
  - Export functionality: Export search results to various formats
  - Advanced filters: Date range, domain, content type filtering
  - Performance metrics: Search time and result quality analytics

  Test Plan for Reviewers

  Quick Verification Steps

  1. Install extension and verify no console errors
  2. Enable history search in options page
  3. Perform search with mixed results (tabs + history)
  4. Test filter UI - click filter button, select different options
  5. Verify click behavior - history opens new tab, tabs get focused
  6. Download debug logs - check log format and content

  Edge Cases to Test

  - No history items - filter should show 0 for history
  - No active tabs - extension handles gracefully
  - Network errors - history fetch failures logged but don't break search
  - Large datasets - performance with 1000+ history items

hunarjain09 committed:
Implement browser history search feature with configurable options and UI enhancements

hunarjain09 committed:
Add filter functionality and logger implementation to popup interface

hunarjain09 committed:
Refactor logging system to use UnifiedLogger across all contexts, enhancing log management and persistence. Update options and popup interfaces to utilize the new logging methods.

hunarjain09 committed:
Enhance result item styling and update message listener parameters for improved readability and functionality

lunu-bounir commented:
Thanks for the PR — great work! Please address the following points:

1. You've submitted a single PR that includes both logger and history changes. Ideally, these should be split: one PR for the logger and another for the history functionality. It is easier to review each one separately.

2. The logger feature should be optional and disabled by default. Once the user enables it from the options page, logging can begin.

3. Both the "history" and "downloads" permissions must be optional. The default selection in the popup interface should be "Active Tabs". If the user switches to "History", prompt for the optional permission (`chrome.permissions.request`). If granted, then include history in the search.

4. I recommend using the standard `<select>` element for the menu instead of a custom implementation. Managing open/close behavior with JavaScript isn’t ideal and lacks accessibility features like keyboard shortcuts and proper focus handling.

5. On the options page, there’s no need for separate enable/disable controls for the history feature. Users can manage this through the popup interface using the select menu. The options page should only contain related settings.

6. I also recommend placing the JavaScript code for history selection in a separate file. It makes review easier.

hunarjain09 commented:
let me break it down to smaller prs addressing all the points 1 at a time. I got little too excited vibe coding this yesterday night. I am sorry for the trouble going through the slob.
